### PR TITLE
New Unit Testing Framework

### PR DESCRIPTION
**Pull request by SmilyOrg**
Old tests have been integrated as LegacyTests. `rake test` fails with error code 1 on any test failure.

### CODE REVIEW

**Comment by bengarney** [comment]
Haha - failed due to:

```
[14:22:04][Execute utility:compileDemo] ===== Compiling UnitTestExample =====
[14:22:04][Execute utility:compileDemo] /Volumes/Swarley/TeamCity/buildAgent/work/160782e93c7262a4/artifacts/lsc
[14:22:04][Execute utility:compileDemo] LSC - JIT Compiler
[14:22:04][Execute utility:compileDemo] Building Main.loom with default settings
[14:22:04][Execute utility:compileDemo] [loom.compiler] Compiling: Main
[14:22:05][Execute utility:compileDemo] ./src/UnitTestExample.ls(3) : error - Failed to import type unittest.TestRunner on line 3
[14:22:05][Execute utility:compileDemo] ./src/UnitTestExample.ls(4) : error - Failed to import type unittest.Assert on line 4
[14:22:05][Execute utility:compileDemo] ./src/UnitTestExample.ls(3) : error - Unable to process type for import unittest.TestRunner
[14:22:05][Execute utility:compileDemo]  in ./src/UnitTestExample.ls
[14:22:05][Execute utility:compileDemo] 
[14:22:05][Execute utility:compileDemo] ./src/UnitTestExample.ls(4) : error - Unable to process type for import unittest.Assert
[14:22:05][Execute utility:compileDemo]  in ./src/UnitTestExample.ls
[14:22:05][Execute utility:compileDemo] 
[14:22:05][Execute utility:compileDemo] ./src/UnitTestExample.ls(19) : error - unable to resolve type for identifier TestRunner
[14:22:05][Execute utility:compileDemo] ./src/UnitTestExample.ls(19) : error - Unable to resolve type for TestRunner
[14:22:05][Execute utility:compileDemo] ./src/UnitTestExample.ls(27) : error - unable to resolve type for identifier Assert
[14:22:05][Execute utility:compileDemo] ./src/UnitTestExample.ls(27) : error - Unable to resolve type for Assert
[14:22:05][Execute utility:compileDemo] 
[14:22:05]
```

So update the example and I think we're good to go! :)

**Comment by SmilyOrg** [comment]
That's weird, because that example totally compiles and works for me (either running it manually or through `rake utility:compileDemos`, updated it just for this reason. Maybe it has something to do with the unittest loomlib not updating?

**Comment by SmilyOrg** [comment]
Okay, found out that it only compiled for me because I ran `rake test` at least once before deploying/updating. That updated the UnitTest loomlib. After adding UnitTest to `Main.build`, it gets picked up and built as well, solving the issue.

I assume this means UnitTest will get included into SDK builds as well, but that's probably what we want, right? It should only include `Assert`, `TestRunner` and `LegacyTest` (to support old-style tests) and none of the actual SDK tests.

**Comment by bengarney** [comment]
Yeah, having a UnitTest.loomlib is ideal.

We also want to make sure that Loom's build box unit tests include the new stuff. But I think you already have that.

**Comment by bengarney** [comment]
OK I think if you rebase/merge master this will work.

**Comment by SmilyOrg** [comment]
Looks green!

File: docs/examples/UnitTestExample/src/UnitTestExample.ls [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    import loom.Application;
+    import system.platform.File;
+    import system.reflection.Assembly;
+    
+    public class UnitTestExample extends Application
+    {
+        [UnitTestHideCall]
+        override public function run():void
+        {
+            //trace(TestRunner.getTests(new ByteArrayTest()));
```
Comment (bengarney) [inline docs/examples/UnitTestExample/src/UnitTestExample.ls]: Can we clean this up? Or clarify/document the options?

File: loom/script/native/lsLuaBridge.h [hunk 2]
```diff
@@ -2679,6 +2679,21 @@ class CallFastSetMember<MemFn, int> : public CallFastMemberBase
     }
 };
 
+template<class MemFn>
+class CallFastSetMember<MemFn, unsigned int> : public CallFastMemberBase
```
Comment (bengarney) [inline loom/script/native/lsLuaBridge.h]: Oh good add. :)

File: sdk/src/test/AllTests.ls [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+*/
+
+package {
+
+    import system.reflection.Type;
+    import unittest.Test;
+    import unittest.TestRunner;
+    import unittest.TestResult;
+    
+    public class AllTests {
```
Comment (bengarney) [inline sdk/src/test/AllTests.ls]: Document purpose of class?

File: sdk/src/unittest/TestRunner.ls [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+            var i:int;
+            var tt:TypeTest;
+            
+            if (typeReport) typeReport.total += typeTests.length;
+            
+            for (i = 0; i < typeTests.length; i++) {
+                tt = typeTests[i];
+                tt.asserts.reset();
+                
+                //IO.write((tt.skip ? "Skipping" : "Running")+" "+tt.type.getFullName()+"   "+(i+1)+" / "+typeTests.length+"\n");
```
Comment (bengarney) [inline sdk/src/unittest/TestRunner.ls]: Dead code?

File: sdk/src/unittest/TestRunner.ls [hunk 5]
```diff
@@ -0,0 +1,379 @@
+package unittest {
+    import system.reflection.Assembly;
+    import system.reflection.MetaInfo;
+    import system.reflection.MethodInfo;
+    import system.reflection.Type;
+    
+    class StatusReport {
```
Comment (bengarney) [inline sdk/src/unittest/TestRunner.ls]: Brief docstrings for these classes?

File: sdk/src/unittest/TestRunner.ls [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+                    tt.skip = type.getMetaInfo("SkipTests") != null;
+                    typeTests.push(tt);
+                }
+            }
+            return typeTests;
+        }
+        
+        [UnitTestHideCall]
+        public static function run(tests:Vector.<Test>, shuffle:Boolean = true) {
+            //var type:Type = target is Type ? target as Type : target.getType();
```
Comment (bengarney) [inline sdk/src/unittest/TestRunner.ls]: Dead code?

File: sdk/src/unittest/TestRunner.ls [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+                //IO.write(lpad(""+passed, " ", 3) + " / " + rpad(""+results.length, " ", 3) + " passed");
+                if (passed < results.length) {
+                    IO.write(" "+String.lpad(""+(results.length-passed), " ", 4)+" failed");
+                    test.results = results;
+                }
+                if (ret != null) IO.write("   "+ret);
+                IO.write("\n");
+            }
+            
+            //IO.write("\n");
```
Comment (bengarney) [inline sdk/src/unittest/TestRunner.ls]: Dead code?

File: sdk/src/unittest/TestRunner.ls [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+                                }
+                            }
+                        }
+                        
+                        IO.write(String.lpad(""+(j+1), " ", 4)+" Assert."+result.info);
+                        if (skip == 0) {
+                            IO.write(" // "+msg);
+                        }
+                        IO.write("\n");
+                        //if (result.message != null) {
```
Comment (bengarney) [inline sdk/src/unittest/TestRunner.ls]: Dead code?

**Comment by bengarney** [comment]
Yes I just nitted it hard! :( I just want to be 100% sure this is solid.

File: sdk/src/unittest/TestRunner.ls (hunk 4 above)
Comment (SmilyOrg) [inline sdk/src/unittest/TestRunner.ls]: Different output styles, not sure which one to use, but the uncommented one is probably fine for now.

**Comment by SmilyOrg** [comment]
Eh, it wasn't nitted that hard :)

**Comment by bengarney** [comment]
Come at me bro

**Comment by bengarney** [comment]
Ah dang, the bytearray pr made this go grey. :( Rebase?

**Comment by bengarney** [comment]
Also please doc the Assert class and add a section to Loom docs showing how to create/run a unit test (assuming loom test works).

**Comment by bengarney** [comment]
This is madsexybeautiful. I am landing it.

**Pull request merged**